nvflare/edge/assessors/model_update.py

```diff
@@ -110,12 +143,52 @@ def process_child_update(self, update: Shareable, fl_ctx: FLContext) -> (bool, O
         with self.update_lock:
             return self._do_child_update(update, fl_ctx)
 
+    def _device_wait_countdown(self, fl_ctx: FLContext):
+        # Check if we've exceeded the timeout
+        if self._check_device_timeout(fl_ctx):
+            usable_devices = set(self.device_manager.available_devices.keys()) - set(
+                self.device_manager.used_devices.keys()
+            )
+            self.log_error(
+                fl_ctx,
+                f"Device wait timeout ({self.device_wait_timeout}s) exceeded. "
+                f"Total devices: {len(self.device_manager.available_devices)}, usable: {len(usable_devices)}, expected: {self.device_manager.device_selection_size}. "
+                f"Device_reuse flag is set to: {self.device_manager.device_reuse}. "
+                "Not enough devices joining, please adjust the server params. Stopping the job.",
+            )
+            # Stop the job due to insufficient devices
+            self.should_stop_job = True
+            # Prepare an empty reply
+            reply = StateUpdateReply(
+                model_version=0,
+                model=None,
+                device_selection_version=self.device_manager.current_selection_version,
+                device_selection=self.device_manager.get_selection(fl_ctx),
+            )
+            return False, reply.to_shareable()
+        else:
+            remaining_time = self._get_remaining_timeout()
+
+            self.log_info(
+                fl_ctx,
+                f"Not enough available devices, waiting for more: " f"Timeout in {remaining_time:.1f} seconds",
+            )
+
     def _do_child_update(self, update: Shareable, fl_ctx: FLContext) -> (bool, Optional[Shareable]):
         report = StateUpdateReport.from_shareable(update)
 
         # Update available devices
         if report.available_devices:
             self.device_manager.update_available_devices(report.available_devices, fl_ctx)
+            # Reset wait timer if we now have enough devices
+            if self.device_wait_start_time is not None and self.device_manager.has_enough_devices(fl_ctx):
+                self.device_wait_start_time = None
+                self.should_stop_job = False  # Reset stop job flag
+                self.log_info(fl_ctx, "Sufficient devices now available, resetting wait timer and stop job flag")
+
+        # Check for device wait timeout if we are waiting for devices
+        if self.device_wait_start_time is not None:
+            self._device_wait_countdown(fl_ctx)
```

Comment:
this "wait_countdown" is only triggered when "_do_child_update" is called, if somehow "_do_child_update" is not called let's say 100 seconds, the device_wait_timeout of 30 seconds is not working as expected.

We might need another thread to handle this device wait timeout.

Make sure to ensure thread safety if state changed: for example: self.should_stop_job